crates/sui-indexer-alt/src/handlers/mod.rs

```diff
@@ -0,0 +1,630 @@
+
+        // Data for checkpoints that haven't been written yet. Note that `pending_rows` includes
+        // rows in `batch`.
+        let mut pending: BTreeMap<u64, Indexed<H>> = BTreeMap::new();
+        let mut pending_rows = 0;
+
+        // Track the high watermark for the pipeline. The pipeline confirms that it has written all
+        // checkpoint data up from the watermark it is initialised with up to and including this
+        // watermark.
+        //
+        // To correctly update the watermark, the committer tracks the watermark it last tried to
+        // write and the watermarks for any checkpoints that have been written since then
+        // ("pre-committed"). After each batch is written, the committer will try to progress the
+        // watermark as much as possible without going over any holes in the sequence of
+        // checkpoints.
+        //
+        // NOTE: When no watermark is provided, it is assumed that the pipeline is starting from
+        // scratch, but we still initialize it as if it is at (after) the genesis checkpoint. This
+        // means we never write a watermark for the genesis checkpoint, but would wait for another
+        // checkpoint to be written out before updating the watermark, which is fine in practice
+        // and simplifies the logic of tracking watermarks.
+        let mut precommitted: BTreeSet<CommitterWatermark<'static>> = BTreeSet::new();
+        let mut watermark =
+            watermark.unwrap_or_else(|| CommitterWatermark::initial(H::NAME.into()));
+
+        // The committer will periodically output a log message at a higher log level to
+        // demonstrate that the pipeline is making progress.
+        let mut next_loud_watermark_update =
+            watermark.checkpoint_hi_inclusive + LOUD_WATERMARK_UPDATE_INTERVAL;
+
+        info!(pipeline = H::NAME, ?watermark, "Starting committer");
+
+        loop {
+            tokio::select! {
+                // Break ties in favour of operations that reduce the size of the buffer.
+                biased;
+
+                _ = cancel.cancelled() => {
+                    info!(pipeline = H::NAME, "Shutdown received, stopping committer");
+                    break;
+                }
+
+                // Time to write out another batch of rows, if there are any.
+                _ = poll.tick(), if pending_rows > 0 => {
+                    let Ok(mut conn) = db.connect().await else {
+                        warn!(pipeline = H::NAME, "Failed to get connection for DB");
+                        cool.reset();
+                        continue;
+                    };
+
+                    let guard = metrics
+                        .committer_gather_latency
+                        .with_label_values(&[H::NAME])
+                        .start_timer();
+
+                    while batch_values.len() < H::CHUNK_SIZE {
+                        let Some(mut entry) = pending.first_entry() else {
```

Comment:
realized that I put the wrong range, updated.

Reply:
This pipeline is only intended for append-only tables, which we can commit to out of order and then update the watermark retroactively (which is handled below), but note that in the example you gave, if `pending` contains `[11, 15)` then it is guaranteed to return `11` first, because `pending` is ordered on checkpoints.